File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.js
/**
 * @param {number[]} costs
 * @param {number} coins
 * @return {number}
 */
var maxIceCream = function(costs, coins) {
    const freq = new Array(100001).fill(0);
    for(const item of costs){
        freq[item]++;
    }
    let ans=0;
    for(let cost=1;cost<=100000;cost++) {
        if(freq[cost]==0) {
            continue;
        }
        const canBuy = Math.min(freq[cost], Math.floor(coins / cost));
        ans += canBuy;
        coins -= canBuy * cost;
        if (coins < cost) break;
    }
    return ans;
};

File: 1833-maximum-ice-cream-bars/1833-maximum-ice-cream-bars.cpp
class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        int ans=0,i=0;
        sort(costs.begin(), costs.end());
        while(i<costs.size()){
            if(coins >= costs[i]){
                ans++;
                coins-=costs[i];
                i++;
            }
            else break;
        }
        return ans;
    }
};